### ISSUE TITLE
[5.8] Adding a built-in self-update command

### ISSUE THREAD

**Issue by RoccoHoward** [laravel-zero/laravel-zero#185]
It would be great to have a self-update command built-in to laravel-zero. I've targetted 5.8 - which may be enough time? Have added this issue on @nunomaduro 's request.

I've developed one for a recent project using a trait/interface to implement it - you can review the code at https://github.com/hnhdigital-os/laravel-console-self-update.

Flexibility is key in providing this particular command as each project will have very different approaches to how/where they store the updated binaries. In a non-laravel-zero project, I store them in a private s3 repo - so using flysystem adapter is an important feature.

The package I've developed has configurability round most of what would be needed and this is currently implemented through set methods.

Changes I'd look at to bring it into the fold:
* Use a config file with environment file (non-master branch or if the project is forked)
* Use guzzle for standard public url access (only using file_get_contents)
* 100% tested
* Provide scripts to generate the files needed (eg versions, hash)

**Comment by nunomaduro** [comment]
I would love to see this: Creating a new component called `self-update-on-build`, under the laravel-zero organization `laravel-zero/self-update-on-build` or even on `hnhdigital-os/self-update-on-build`.

1. The component should be available on the `app:install` command. Providing a new component is as easy as: https://github.com/laravel-zero/framework/pull/292/files#diff-57ffd60d2a32bd7a4bdbdf1e08e47147R42.
2. After installing the new component, and after making an `app:build` the command `self-update` will be available on the built app only.

Notes: By default, the behavior of the `self-update` would be search on the `https://github.com/project/name/tree/vX.X.X` for the built name. But there would be an interface for coding different approachs. Tell me what do you think.

**Comment by RoccoHoward** [comment]
GraphQL API v4 query for latest version:-

```
{
  repository(owner: "laravel-zero", name: "laravel-zero") {
    refs(refPrefix: "refs/tags/", first: 1, orderBy: {field: TAG_COMMIT_DATE, direction: DESC}) {
      edges {
        node {
          name
          target {
            oid
            ... on Tag {
              commitUrl
              tagger {
                date
              }
            }
          }
        }
      }
    }
  }
}
```

**Comment by nunomaduro** [comment]
> The package name is a bit odd though, wouldn't it be better as self-update-command?

Yes. You are right.

> The default location of the file to download would actually be:-https://github.com/OWNER/PROJECT/raw/master/download/TAG/BUILD_NAME.

Careful here. I think the link should be: `https://github.com/OWNER/PROJECT/tree/TAG/builds/BUILD_NAME`

> Getting the BUILD_NAME would be config('app.name')?

The build name, by default, is the app name yes. Let's keep it simple, and default to the app name.

> Setting the default URL requires knowing the details to generate the default URL - the project owner/name would need to be set via env or directly in a config file (unless you have a way to automate that?)

You can extract that from the `composer.json`.

> Getting the TAG for latest tag is possible through GitHub GraphQL API v4 - I'd need to test it against a few repo's to confirm. Doing it via GraphQL let's you order the returned versions to the latest and one record. The only issue is it requires being authenticated - so a token would need to be set in config.

Let's think about this later.

PS: Thank you for your hard work on this. Credits will be propertly assigned.

**Comment by RoccoHoward** [comment]
> Careful here. I think the link should be: https://github.com/OWNER/PROJECT/tree/TAG/builds/BUILD_NAME
 
Took me a bit to get what you were saying. Got it now. So used to putting binaries in a different branch/folder based/non tagged. Only have to add ?raw=true to the end of the URL to make it work. I'll be testing it against a test repo anyway.

> You can extract that from the composer.json.

Of course. Just need to be clear to anyone that this needs to be updated - when you start a new project - this is all laravel-zero details in the composer.json file which is why potentially making it config based forces the author to put the right details in.

> PS: Thank you for your hard work on this. Credits will be properly assigned.

No problem. It's good to contribute back in some way.

**Comment by nunomaduro** [comment]
Good to know that we are on the same page. Just hit my back when you have a POC of this.

**Comment by mrdionjr** [comment]
Great! I was going to create the same issue. How is it going? You can check out https://github.com/humbug/phar-updater if you need some ideas.

**Comment by nunomaduro** [comment]
@RoccoHoward Have you checked phar-updater? It's also used in this Laravel Zero app: https://github.com/linuxjuggler/cloudflare-cli/blob/master/app/Commands/SelfUpdateCommand.php.

**Comment by RoccoHoward** [comment]
@nunomaduro  I'm looking at it over the weekend - will let you know if it does the trick. Might save some time and we won't have to maintain much.

**Comment by nunomaduro** [comment]
@RoccoHoward Have news on this?

**Comment by RoccoHoward** [comment]
@nunomaduro Sorry. I work for myself and my time has been very limited with kids back at school + wife covering her boss + paid work and business trip.

Likely be able to come back to this in 3-4 weeks time - sorry for missing the 5.8 window.

**Comment by nunomaduro** [comment]
No worries. Maybe @ijpatricio wants to take this? Or someone else?

**Comment by linuxjuggler** [comment]
Hey @nunomaduro, Can you have a look at https://github.com/linuxjuggler/framework/commit/d03df861aeb160dbd49afe9e81167c06bc447bb3 .

Please note that this is for sure *not a full solution* its just a simple implementation which needs a lot of enhancement.

let me know what you think

**Comment by nunomaduro** [comment]
@linuxjuggler Can you do a pull request against develop?

**Comment by linuxjuggler** [comment]
sure thing @nunomaduro but just a note, I didnt test it yet so we should look at it carefully

**Comment by linuxjuggler** [comment]
Also, I was thinking about the whole `app:install` feature and wondering if we can just separate it into a separate package so it would be easier to build packages without touching the framework.

This can be for another topic, which am happy to discuss with you later @nunomaduro

**Comment by nunomaduro** [comment]
This feature was been included on Laravel Zero 5.8.
